File: leetcode/[0094]binary-tree-inorder-traversal.rs
// https://leetcode.com/problems/binary-tree-inorder-traversal/

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        return solution2(root);
    }
}



// recursive: 0 ms
fn solution1(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut a = Vec::<i32>::new();
    let mut root = root;

    rec(&root, &mut a);

    return a;
}
fn rec(n: &Option<Rc<RefCell<TreeNode>>>, a: &mut Vec<i32>) {
    if n.is_some() {
        rec(&n.as_ref().unwrap().borrow().left, a);
        a.push(n.as_ref().unwrap().borrow().val);
        rec(&n.as_ref().unwrap().borrow().right, a);
    }
}



// iterative: 0 ms
fn solution2(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    if !root.is_some() {
        return vec![];
    }
    
    let mut a = Vec::<i32>::new();
    let mut st = vec![(root.clone(), false)];

    while st.len() > 0 {
        let (n, v) = st.pop().unwrap();

        if n.is_some() {
            if v {
                a.push(n.as_ref().unwrap().borrow().val);
            } else {
                st.push((n.as_ref().unwrap().borrow().right.clone(), false));
                st.push((n.clone(), true));
                st.push((n.as_ref().unwrap().borrow().left.clone(), false));
            }
        }
    }

    return a;
}

File: leetcode/[0094]binary-tree-inorder-traversal.ts
// https://leetcode.com/problems/binary-tree-inorder-traversal/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function inorderTraversal(root: TreeNode | null): number[] {
    return solution3(root);
}



// iterative: 0 ms
function solution1(root: TreeNode | null): number[] {
    if (!root) return [];

    const a = [];
    const st = [[root, false]] as [TreeNode, boolean][];

    while (st.length > 0) {
        const [n, v] = st.pop()!;

        if (n) {
            if (v) {
                a.push(n.val);
            } else {
                st.push(
                    [n.right!, false], [n, true], [n.left!, false]
                );
            }
        }
    }

    return a;
}



// recursive: 0 ms
function solution2(root: TreeNode | null): number[] {
    const a = [] as number[];

    const f = function (n: TreeNode | null) {
        if (n) {
            f(n.left);
            a.push(n.val);
            f(n.right);
        }
    };

    f(root);

    return a;
}



// recursive2: 0 ms
function solution3(root: TreeNode | null): number[] {
    return (root) ? [...solution3(root.left), root.val, ...solution3(root.right)] : [];
}
